sdk/inc/azure/core/az_context.h

```diff
@@ -16,6 +16,7 @@
 #define _az_CONTEXT_H
 
 #include <azure/core/az_result.h>
+#include <azure/core/internal/az_precondition_internal.h>
```

Comment:
A public header must not include an internal header. Make the methods that need preconditions non-inlined and move them to the `.c` source file.

Reply:
That seems an arbitrary requirement.  We should move the preconditions so they can be used in headers

Reply:
> That seems an arbitrary requirement.

I don't follow. We don't want to make internal APIs/types accidentally visible to the caller.
https://azure.github.io/azure-sdk/clang_implementation.html#coding-style
⛔️ DO NOT include internal or private headers in public headers.

> We should move the preconditions so they can be used in headers

All these precondition macros are internal. Are you suggesting we make them public?

Reply:
Preventing the types from being used in the headers means we can't inline the functions.  Also these are _az functions  which by definition are internal.  They might be exposed to the user but they are documented as such.

Reply:
There's a few things to unpack here.

> Preventing the types from being used in the headers means we can't inline the functions

1) If it is an important perf-sensitive code path, especially used within tight loops, and wouldn't otherwise be inlined by the compiler automatically, without the hint, it is possible to inline the function by following what we do in `az_span_create`. Feel free to follow that approach.
https://github.com/Azure/azure-sdk-for-c/blob/9137165c0309e379aa4a3ecdb032d99e38380df9/sdk/inc/azure/core/az_span.h#L67-L76
https://github.com/Azure/azure-sdk-for-c/blob/9137165c0309e379aa4a3ecdb032d99e38380df9/sdk/src/azure/core/az_span.c#L23-L36

> Also these are _az functions which by definition are internal. They might be exposed to the user but they are documented as such.

2) We do not expose any such internal methods to the public via public header files. Any non-public `_az` function is either static/within the source .c file, OR in an internal/private header file if they are used across various files and component. Those header files are not exposed from any of the public header files either and are not used in samples. Why would we want to make an exception to that rule here?

If a user includes `az_span.h` and then sees/gets access to some internal method that started with `_az`, which they accidentally call (without realizing it was internal), it makes changing or breaking those methods, which are implementation detail, more difficult. Now, if the user really wanted to, they could explicitly include some internal header file but that is a more explicit, intentional action.

Also, although we should fix this, keep in mind, that not all of our internal and private methods start with `_az`. Without some static linting tool/rules, as part of code maintainability and code reviews, it is a lot easier to see what includes are in a header file and catch bugs than making sure no internal API is accidentally exposed.
https://github.com/Azure/azure-sdk-for-c/blob/9137165c0309e379aa4a3ecdb032d99e38380df9/sdk/inc/azure/core/internal/az_http_internal.h#L98
https://github.com/Azure/azure-sdk-for-c/blob/9137165c0309e379aa4a3ecdb032d99e38380df9/sdk/inc/azure/core/internal/az_http_internal.h#L180

3) In this particular instance, I don't think it's necessary to mark these methods inlined anyway. They are not perf-sensitive, won't be called often (certainly not in loops, and they might evolve overtime to become more complicated, making inlining not profitable. Generally, we should only inline APIs that we know are relatively small and customers would benefit from them being inlined. Keep in mind, inlining is just a hint to the compiler, and it is free to ignore it anyway. Even if the precondition macros were public, I would recommend not inlining. To get the precondition benefit, we had some APIs that are generally more perf sensitive and called in potentially hot code paths, that we still didn't inline. For example, the JSON init methods, because they are still only called once and not in tight loops, and bulk of the work/time is spent in the actual processing.
https://github.com/Azure/azure-sdk-for-c/blob/9137165c0309e379aa4a3ecdb032d99e38380df9/sdk/src/azure/core/az_json_reader.c#L15-L20

**All that said, if you feel strongly about this, please bring this up with the rest of the team, lets discuss this, and change the guidelines, or make the precondition macros public.**

Reply:
I did look at what span did b/c I couldn't believe I was the first to hit this.  I'm against the span solution because of the code duplication.  There is a risk the fix in the C file may not make it into the .h inline version or vice-versa.  Is there performance data that shows the duplicated code is worth the overhead of maintaining two functions?